File: app/src/main/java/com/hjc/reader/ui/search/child/SearchResultFragment.java
package com.hjc.reader.ui.search.child;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.hjc.baselib.event.EventManager;
import com.hjc.baselib.event.MessageEvent;
import com.hjc.baselib.fragment.BaseMvmFragment;
import com.hjc.reader.R;
import com.hjc.reader.bean.response.WanListBean;
import com.hjc.reader.constant.EventCode;
import com.hjc.reader.constant.RoutePath;
import com.hjc.reader.databinding.FragmentSearchResultBinding;
import com.hjc.reader.ui.search.adapter.SearchAdapter;
import com.hjc.reader.utils.helper.AccountManager;
import com.hjc.reader.utils.helper.RouteManager;
import com.hjc.reader.viewmodel.search.SearchResultViewModel;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.listener.OnRefreshLoadMoreListener;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.List;

/**
 * @Author: HJC
 * @Date: 2020/9/11 10:55
 * @Description: 搜索结果页面
 */
public class SearchResultFragment extends BaseMvmFragment<FragmentSearchResultBinding, SearchResultViewModel> {

    private SearchAdapter mAdapter;

    private int mPage = 0;

    private String mKeyword;

    public static SearchResultFragment newInstance() {
        return new SearchResultFragment();
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_search_result;
    }

    @Override
    protected SearchResultViewModel getViewModel() {
        return new ViewModelProvider(this).get(SearchResultViewModel.class);
    }

    @Override
    protected int getBindingVariable() {
        return 0;
    }

    @Override
    protected void initView() {
        setLoadSir(mBindingView.smartRefreshLayout);

        LinearLayoutManager manager = new LinearLayoutManager(mContext);
        mBindingView.rvList.setLayoutManager(manager);

        mAdapter = new SearchAdapter();
        mBindingView.rvList.setAdapter(mAdapter);

        mAdapter.setAnimationWithDefault(BaseQuickAdapter.AnimationType.ScaleIn);
    }

    @Override
    protected void initData(@Nullable Bundle savedInstanceState) {
        EventManager.register(this);
    }

    /**
     * 点击搜索后的逻辑
     */
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    public void handlerEvent(MessageEvent<String> messageEvent) {
        if (messageEvent.getCode() == EventCode.SEARCH_RESULT) {
            mKeyword = messageEvent.getData();
            mPage = 0;
            mBindingView.rvList.scrollToPosition(0);
            mViewModel.loadSearchData(mPage, mKeyword, true);
        }
    }

    @Override
    protected void observeLiveData() {
        mViewModel.getSearchLiveData().observe(this, data -> {
            mBindingView.smartRefreshLayout.finishRefresh();
            mBindingView.smartRefreshLayout.finishLoadMore();

            if (data != null) {
                if (mPage == 0) {
                    mAdapter.setNewInstance(data);
                } else {
                    mAdapter.addData(data);
                }
            }
        });

        mViewModel.getArticleLiveData().observe(this, data -> {
            mAdapter.notifyDataSetChanged();
        });
    }

    @Override
    protected void addListeners() {
        mBindingView.smartRefreshLayout.setOnRefreshLoadMoreListener(new OnRefreshLoadMoreListener() {
            @Override
            public void onLoadMore(@NonNull RefreshLayout refreshLayout) {
                mPage++;
                mViewModel.loadSearchData(mPage, mKeyword, false);
            }

            @Override
            public void onRefresh(@NonNull RefreshLayout refreshLayout) {
                mPage = 0;
                mViewModel.loadSearchData(mPage, mKeyword, false);
            }
        });

        mAdapter.setOnItemClickListener((adapter, view, position) -> {
            List<WanListBean.DataBean.DatasBean> dataList = mAdapter.getData();
            WanListBean.DataBean.DatasBean bean = dataList.get(position);
            RouteManager.jumpToWeb(bean.getTitle(), bean.getLink());
        });

        mAdapter.addChildClickViewIds(R.id.cb_collect);
        mAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            if (view.getId() == R.id.cb_collect) {
                List<WanListBean.DataBean.DatasBean> dataList = mAdapter.getData();
                WanListBean.DataBean.DatasBean bean = dataList.get(position);

                //收藏或取消收藏前判断是否已登录
                if (AccountManager.getInstance().isLogin()) {
                    if (!bean.isCollect()) {
                        mViewModel.collectArticle(bean);
                    } else {
                        mViewModel.unCollectArticle(bean);
                    }
                } else {
                    bean.setCollect(false);
                    adapter.notifyItemChanged(position);
                    RouteManager.jump(RoutePath.URL_LOGIN);
                }
            }
        });
    }

    @Override
    protected void onRetryBtnClick(View v) {
        mPage = 0;
        mViewModel.loadSearchData(mPage, mKeyword, true);
    }

    @Override
    protected void onSingleClick(View v) {

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventManager.unregister(this);
    }
}

File: module-main/src/main/java/com/hjc/learn/main/ui/search/child/SearchResultFragment.kt
package com.hjc.learn.main.ui.search.child

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.alibaba.android.arouter.facade.annotation.Route
import com.chad.library.adapter.base.BaseQuickAdapter.AnimationType
import com.hjc.learn.main.R
import com.hjc.learn.main.adapter.SearchAdapter
import com.hjc.learn.main.databinding.MainFragmentSearchResultBinding
import com.hjc.learn.main.viewmodel.search.SearchResultViewModel
import com.hjc.library_base.fragment.BaseFragment
import com.hjc.library_common.event.EventManager
import com.hjc.library_common.event.MessageEvent
import com.hjc.library_common.global.EventCode
import com.hjc.library_common.router.RouteManager
import com.hjc.library_common.router.path.RouteLoginPath
import com.hjc.library_common.router.path.RouteMainPath
import com.hjc.library_net.utils.AccountHelper
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnRefreshLoadMoreListener
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/**
 * @Author: HJC
 * @Date: 2020/9/11 10:55
 * @Description: 搜索结果页面
 */
@Route(path = RouteMainPath.URL_FRAGMENT_SEARCH_RESULT)
class SearchResultFragment : BaseFragment<MainFragmentSearchResultBinding, SearchResultViewModel>() {

    private lateinit var mAdapter: SearchAdapter

    private var mPage = 0
    private var mKeyword: String? = null

    override fun getLayoutId(): Int {
        return R.layout.main_fragment_search_result
    }

    override fun createViewModel(): SearchResultViewModel {
        return ViewModelProvider(this)[SearchResultViewModel::class.java]
    }

    override fun initView() {
        super.initView()

        initLoadSir(mBindingView.smartRefreshLayout)

        val manager = LinearLayoutManager(mContext)
        mBindingView.rvList.layoutManager = manager

        mAdapter = SearchAdapter()
        mBindingView.rvList.adapter = mAdapter

        mAdapter.setAnimationWithDefault(AnimationType.ScaleIn)
    }

    override fun initData(savedInstanceState: Bundle?) {
        EventManager.register(this)
    }

    /**
     * 点击搜索后的逻辑
     */
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    fun receiveEvent(messageEvent: MessageEvent<String?>) {
        if (messageEvent.getCode() === EventCode.SEARCH_RESULT_CODE) {
            mKeyword = messageEvent.getData()
            mPage = 0
            mBindingView.rvList.scrollToPosition(0)
            mViewModel?.loadSearchData(mPage, mKeyword, true)
        }
    }

    override fun observeLiveData() {
        super.observeLiveData()

        mViewModel?.run {
            searchLiveData.observe(this@SearchResultFragment) { result ->
                if (mPage == 0) {
                    mAdapter.setNewInstance(result)
                } else {
                    mAdapter.addData(result)
                }
            }

            refreshData.observe(this@SearchResultFragment) { result ->
                if (result) {
                    mBindingView.smartRefreshLayout.finishRefresh()
                    mBindingView.smartRefreshLayout.finishLoadMore()
                }
            }

            articleLiveData.observe(this@SearchResultFragment) { mAdapter.notifyItemChanged(it.position) }
        }
    }

    override fun addListeners() {
        mBindingView.smartRefreshLayout.setOnRefreshLoadMoreListener(object : OnRefreshLoadMoreListener {
            override fun onRefresh(refreshLayout: RefreshLayout) {
                mPage = 0
                mViewModel?.loadSearchData(mPage, mKeyword, false)
            }

            override fun onLoadMore(refreshLayout: RefreshLayout) {
                mPage++
                mViewModel?.loadSearchData(mPage, mKeyword, false)
            }

        })

        mAdapter.setOnItemClickListener { _, _, position ->
            val dataList = mAdapter.data
            val bean = dataList[position]
            RouteManager.jumpToWeb(bean.title, bean.link)
        }

        mAdapter.addChildClickViewIds(R.id.cb_collect)
        mAdapter.setOnItemChildClickListener { adapter, view, position ->
            when (view.id) {
                R.id.cb_collect -> {
                    val dataList = mAdapter.data
                    val bean = dataList[position]

                    //收藏或取消收藏前判断是否已登录
                    if (AccountHelper.isLogin) {
                        if (!bean.collect) {
                            mViewModel?.collectArticle(bean, position)
                        } else {
                            mViewModel?.unCollectArticle(bean, position)
                        }
                    } else {
                        bean.collect = false
                        adapter.notifyItemChanged(position)
                        RouteManager.jump(RouteLoginPath.URL_LOGIN)
                    }
                }
            }
        }
    }

    override fun onSingleClick(v: View?) {

    }

    override fun onRetryBtnClick(v: View?) {
        super.onRetryBtnClick(v)
        mPage = 0
        mViewModel?.loadSearchData(mPage, mKeyword, true)
    }


    override fun onDestroy() {
        super.onDestroy()
        EventManager.unregister(this)
    }
}